src/js/common/validator.js

```diff
@@ -1,122 +1,470 @@
+
+        else {
+            return rawSegments;
         }
-    },
-    listeners: {
-        "onCreate.init": {
-            funcName: "gpii.schema.validator.ajv.init",
-            args:     ["{that}"]
-        },
-        "onSchemasLoaded.refreshSchemas": {
-            funcName:      "gpii.schema.validator.ajv.refreshSchemas",
-            args:          ["{that}"]
+    };
+
+    /**
+     *
+     * Trim a leading dot if found.  Required because the `missingProperty` field used by AJV is sometimes (but not always) preceded by a leading dot.
+     *
+     * NOTE: This function is a non-API function, i.e. one that assists public functions in performing their work, but
+     * which is not guaranteed to remain available.
+     *
+     * @param {String} rawSegment - The string to be sanitised.
+     * @return {String} - The string, with a leading dot removed.
+     */
+    gpii.schema.validator.trimLeadingDot = function (rawSegment) {
+        return typeof rawSegment === "string" && rawSegment.indexOf(".") === 0 ? rawSegment.substring(1) : rawSegment;
+    };
+
+    /**
+     *
+     * Convert a JSON Pointer (https://tools.ietf.org/html/rfc6901) to a set of EL Path segments that can be used with
+     * `fluid.get`.
+     *
+     * @param {String} jsonPointer - A JSON pointer.
+     * @return {Array<String>} - An array of strings representing the path to the same location as EL path segments.
+     *
+     */
+    gpii.schema.validator.jsonPointerToElPath = function (jsonPointer) {
+        // Discard the leading portion of the URL content if it's found.
+        var rawSegments = jsonPointer.substring(jsonPointer.indexOf("#") + 1).split("/").filter(gpii.schema.removeEmptyItems);
+        // Handle slash escaping, 'this~1that` => `this/that` and tilde escaping, `topsy~0turvy` => `topsy~turvy`
+        var unescapedSegments = fluid.transform(rawSegments, function (segment) {
+            return segment.replace("~1", "/").replace("~0", "~");
+        });
+        return unescapedSegments;
+    };
+
+    /**
+     *
+     * Examine an EL path to a failing rule within a GSS schema, looking for help in presenting a cleaner error
+     * message.  Error definitions look like:
+     *
+     * {
+     *  "short": {
+     *      "type": "string",
+     *      "errors": "short-invalid-generic-message-key"
+     *  },
+     *  "long": {
+     *      "format": "long-invalid-format-invalid-key",
+     *      "required": "long-required-key",
+     *      "": "long-invalid-key"
+     *  }
+     * }
+     *
+     * When using the "long" notation, you can still fail over to a generic message using the key `""`, as shown above.
+     *
+     * NOTE: This function is a non-API function, i.e. one that assists public functions in performing their work, but
+     * which is not guaranteed to remain available.
+     *
+     * @param {Array<String>} rulePath - An array of EL path segments.
+     * @param {Object} gssSchema - A GSS schema.
```

Comment:
Is there anything sensible we can do with a @typedef in order to describe what could go in this object? I imagine in time we could make a "JSON schema to eslint @typedef compiler" that could generate such docs from hints and rules.

Reply:
@amb26, I have doubts that we can do represent the full range of what's possible with GSS and/or JSON schemas in JSDoc typedefs.  Up front, some things do seem doable slash easy:

1.  We can represent deep @property entries using a very familiar dot notation.
2. Optionality is trivial to represent, i.e. square braces.
3. Type mapping of JSON schema primitives should be straightforward.
4. Recursion (which we need for things like the "schema" field in each setting entry) seems to be supported, although some things that consume JSDocs [don't always handle it well](https://github.com/Microsoft/TypeScript/issues/15597).

I'm a bit more concerned about:

1. Usages of `oneOf`, i.e. "must match only one of these schemas".  Anything more complex than "{String|Array<String>}" is likely to be difficult to express unless we want to keep track of and generate typedefs for each allowed sub-type.
2. Usages of `allOf`, i.e. "must match all of these schemas".  This will involve merging material ourselves to detect the full range of rules and sub-properties.
3. Usages of `anyOf`, i.e. "must match one or more of these schemas".  If we have 3 possibilities, we might need to express 6 sub-typedefs before we can safely "or" them together in JSDoc syntax.

Just to organise my thinking and give me better specific examples, I plan to sketch out a typedef for GSS.  If it proves too difficult, a link to the documentation seems appropriate.

Reply:
@amb26, I have committed a first pass at this.  If you have even a little time to chat tomorrow we could discuss further defanging some of the complex bits of the JSON Schema language, writing everything out got me thinking.

Reply:
I think the first pass is more than good enough to be going along with, given the purpose of these @typedefs for a very long time will just be to aid orientation and reader comprehension. Just remember to actually change the @param type here to @GssSchema